### DISCUSSION TITLE
some OAL and PromQL API questions

### DISCUSSION THREAD

**Discussion by unihon** [apache/skywalking#12164] [Q&A]
1. how to count the sum of the http response status quantities of service or endpoint
eg metrics:
http_status_code_count{service="xx", endpoint="yy"}
data:
xx,yy count 200 = 10
xx,yy count 404 = 5

1. how query endpoint metrics with specific instances
eg metrics:
endpoint_sla{service="xx", endpoint="yy", instance="zz"}
endpoint_cpm{service="xx", endpoint="yy", instance="zz"}

**Comment by wu-sheng** [comment]
> 1. how to count the sum of the http response status quantities of service or endpoint

OAL has `sum` function, you could refer to those docs. Don't count on us to write OAL for you. That is not how things work.

> 2. how query endpoint metrics with specific instances

Not supported. Endpoint is service level, not instance level.

**Comment by unihon** [comment]
![image](https://github.com/apache/skywalking/assets/42078581/e9c89535-5ad5-470f-a8de-b01578122894)

Does OAL Filter support "not in"?

**Reply by wu-sheng** [reply]
There is `not contain`, no `not it`. Docs are correct.

**Comment by unihon** [comment]
```
service_code_200 = from(Service.*).filter(httpResponseStatusCode == 200).count()
service_code_302 = from(Service.*).filter(httpResponseStatusCode == 302).count()
service_code_400 = from(Service.*).filter(httpResponseStatusCode == 400).count()
service_code_401 = from(Service.*).filter(httpResponseStatusCode == 401).count()
service_code_403 = from(Service.*).filter(httpResponseStatusCode == 403).count()
service_code_404 = from(Service.*).filter(httpResponseStatusCode == 404).count()
...
service_code_500 = from(Service.*).filter(httpResponseStatusCode == 500).count()
...
```
I can currently count the status codes of each status code through the above methods. Is there any other recommended method?

**Reply by wu-sheng** [reply]
If you want to do that, here is the right way. Or, you could implement a new function as a labeled value function to count. Upstream doesn't have that.

**Comment by unihon** [comment]
definition of metrics
```
service_sla = from(Service.*).percent(status == true);
service_http_status_code_500 = from(Service.*).filter(httpResponseStatusCode == 500).count()
```

![image](https://github.com/apache/skywalking/assets/42078581/aa1993f2-6ec1-41ab-b2e2-2faab9f521ff)

![image](https://github.com/apache/skywalking/assets/42078581/d70c298d-f696-442b-81cb-3219ac8a1f95)

Under what circumstances is the value of `Service.status` equal to `true` and under what circumstances is it `false`?
From my understanding, http status code 500, `Service.status` is false, sla will be less than 100%.

**Reply by unihon** [reply]
skywalking ui:
![image](https://github.com/apache/skywalking/assets/42078581/cf1365b8-52f7-4089-a025-c803901d81ff)

**Reply by wu-sheng** [reply]
> Under what circumstances is the value of Service.status equal to true and under what circumstances is it false?

That status is controlled by the span in the tracing case, which means, this is plugin implementation's responsibility to declare this. Meanwhile, status code oriented analysis is based on OAL, which is dynamic and mutable.

**Reply by wu-sheng** [reply]
Usually, HTTP response code >= 400 would be considered as error status in most plugins.

**Comment by hasaiki123** [comment]
It means any functions with PromQL in Prometheus cannot be supported in Skywalking PromQL？

**Reply by hasaiki123** [reply]
If we want to use like sum、avg、irate... It does not work。
